jbmc/unit/java-testing-utils/require_goto_statements.cpp

```diff
@@ -11,11 +11,47 @@ Author: Diffblue Ltd.
 #include <testing-utils/catch.hpp>
 
 #include <algorithm>
-#include <util/expr_iterator.h>
 #include <goto-programs/goto_functions.h>
+#include <goto-programs/show_goto_functions.h>
+#include <iostream>
 #include <java_bytecode/java_types.h>
+#include <util/expr_iterator.h>
+#include <util/expr_util.h>
 #include <util/suffix.h>
 
+/// Given an expression, attempt to find the underlying symbol it represents
+/// by skipping over type casts and removing balanced dereference/address_of
+/// operations
+optionalt<symbol_exprt>
+root_object(const exprt &lhs_expr, const symbol_tablet &symbol_table)
```

Comment:
As said in the comment, there is `object_descriptor_exprt`. I think it may be necessary to extend `object_descriptor_exprt::build_object_descriptor_rec` with code handling paired `dereference_exprt/address_of_exprt`, but then it should do the trick for you. I believe all the code that this needs in addition is (some variant of):
```
else if(expr.id() == ID_dereference)
{
  const dereference_exprt &deref = to_dereference_expr(expr);
  if(deref.pointer().id() == ID_address_of)
  {
    const address_of_exprt &address_of = to_address_of_expr(deref.pointer());
    dest.object() = address_of.object();
    build_object_descriptor_rec(ns, address_of.object(), dest);
  }
}
```

Reply:
Yes I saw the `object_descriptor_exprt` and saw that it didn't handle the ref/deref. In the interest of not wanting to break anything else that uses this, I will attempt this refactor in a follow up PR too if that is acceptable to you.

Reply:
Sure!

Reply:
I'll post a PR in a few minutes.

Reply:
See #3898.